### PR TITLE
fix(mme): PDUSessionResourceSetupRequest failure randomly

### PR DESCRIPTION
**Pull request by ganeshg87**
Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

## Summary
  fix(mme): PDUSessionResourceSetupRequest failure randomly

## Test Plan

 below flows verified using UERANSIM
      1. registration setup
      2. pdu session
      3. deregistration

## Additional Information

1. pdu session resource setup request failed due to number active pdusession variable is uninitialized 
bb87a45ae1c55486f51997eaa008cf6f46ae3f6f220e88b4b145572bc333e30f
2. pdu session resource setup request failed due to pds_id.guti  variable is uninitilized 
image

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp [hunk 1]
```diff
@@ -1191,11 +1191,6 @@ uint16_t amf_as_establish_cnf(
   // Setting-up the AS message
   as_msg->ue_id = msg->ue_id;
 
-  if (msg->pds_id.guti == NULL) {
-    OAILOG_WARNING(LOG_NAS_AMF, "AMFAS-SAP - GUTI is NULL...");
-    OAILOG_FUNC_RETURN(LOG_NAS_AMF, ret_val);
-  }
-
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: why was this check here before and why is it removed now?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp [hunk 2]
```diff
@@ -139,6 +139,7 @@ ue_m5gmm_context_s* amf_create_new_ue_context(void) {
     OAILOG_ERROR(LOG_AMF_APP, "Failed to allocate memory for UE context \n");
     return NULL;
   }
+  memset(new_p, 0, sizeof(ue_m5gmm_context_s));
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp]: add unit test coverage for `amf_create_new_ue_context`

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp [hunk 3]
```diff
@@ -217,6 +218,7 @@ ue_m5gmm_context_s* amf_ue_context_exists_amf_ue_ngap_id(
 smf_context_t* amf_insert_smf_context(
     ue_m5gmm_context_s* ue_context, uint8_t pdu_session_id) {
   smf_context_t smf_context;
+  memset(&smf_context, 0, sizeof(smf_context));
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp]: add unit test coverage for `amf_insert_amf_context`

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp (hunk 1 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: The above code was committed from Chandra on April 12(#5984).
reviewed PR, above specific variable, is not initialized.

when I was using the UERANSIM script with Registration & PDUSession, sometimes it was failing due to the above check.
backtraced pds_id.guti variable is unused & uninitilzed value.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 2 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp]: there was no issue in  ue_m5gmm_context_s initialization, safer side I added.
but ue_m5gmm_context_s  struct   allocated with c++ constructor(new ue_m5gmm_context_s()  ),properly all members are initialized, removed memset and validated with UT.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp (hunk 3 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context.cpp]: comment addressed added test coverage.

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp (hunk 1 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: till now we did not see any issue because pds_id.guti value garbage, and amf not used this variable.

Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: basically, amf_sap structure is not initialized,  I have called memset all places where  amf_sap  defined

Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: function local variables initialization is a basic coding guideline, it was missed in base code.
I did not change any functionality and not added any UT cases.
 
Please provide your feedback,

**Comment by ganeshg87** [comment]
- basic sanity done with UERANSIM
           initial registration, pdusesssion, deregister flows.
image
[mme_ue_register_06Aug2021_092923.log](https://github.com/magma/magma/files/6944660/mme_ue_register_06Aug2021_092923.log)

- added new UT cases and executed UT modules

image

**Comment by ganeshg87** [comment]
I have replaced memset with C++11 '{}' as per Yogesh review comment.

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 4]
```diff
@@ -10,8 +10,28 @@
...
+
+task_zmq_ctx_t grpc_service_task_zmq_ctx;
+
+const task_info_t tasks_info[] = {
+    {THREAD_NULL, "TASK_UNKNOWN", "ipc://IPC_TASK_UNKNOWN"},
+#define TASK_DEF(tHREADiD)                                                     \
+  {THREAD_##tHREADiD, #tHREADiD, "ipc://IPC_" #tHREADiD},
+#include <tasks_def.h>
+#undef TASK_DEF
+};
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp]: I do not see any itti messages or tasks being used in these tests. What is the reason for including these?

Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: addressed review comment removed unused itti header files.

**Comment by ganeshg87** [comment]
- verified oai UT cases
image

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 5]
```diff
@@ -10,11 +10,15 @@
...
 #include "util_nas5g_pkt.h"
-#include "dynamic_memory_check.h"
 #include <gtest/gtest.h>
+#include "intertask_interface.h"
+#include "../../tasks/amf/amf_app_ue_context_and_proc.h"
+
 
 using ::testing::Test;
 
+task_zmq_ctx_t grpc_service_task_zmq_ctx;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp:20]: is this used? If not remove it.

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 6]
```diff
@@ -10,11 +10,15 @@
  */
 
 #include "util_nas5g_pkt.h"
-#include "dynamic_memory_check.h"
 #include <gtest/gtest.h>
+#include "intertask_interface.h"
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp:14]: is this header used? If not remove it.

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 5 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp:20]: n11 has a dependency with grpc_service, it requires  grpc_service_task_zmq_ctx, If I remove the above, getting the below error due to dependency.
image

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 6 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp:14]: grpc_service_task_zmq_ctx was declared in intertask_interface.h

**Review by ardzoht — APPROVED** [review]
lgtm

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**